### ISSUE TITLE
Support Exceptions with status codes != 500

### ISSUE THREAD

**Issue by flowdy** [kraih/mojo#1014]
* Mojolicious version: 6.53
* Perl version: 5.20
* Operating system: Debian GNU/Linux 8.3

### Steps to reproduce the behavior

        $c->reply->not_found(); # status 404, but error message too unspecific
        $c->reply->exception("Where've all the flowers gone?"); # status code 500 vs. 404

### Expected behavior

In addition to reply->exception, I'd need a broader way to specify how to handle/render instances of my custom exception classes. Especially they have got an attribute http_status, but currently it is  void. There should be a way to set a status code even when exceptions are thrown. I expect to be able to overwrite the status with the appropriate one in before_render hook.

Yet better an approach would be to provide a render_exception hook to activate when an exception object is thrown. It is meant to return args for render.

    $self->hook( render_exception => sub {
        my @render_args;
        if ( $_->isa('MyApp::Error') ) {
            @render_args = (
               status => $_->http_status, message => $_->message,
               error => ref $_,
             );
        }
        return $c->stash("is_restapi_req") ? json => { @render_args }
             :                               @render_args
             ;
    });

### Actual behavior

reply->not_found sets the status to 404 which is correct. However, the content message cannot be specific. One can use reply->exception(), but the status code would be 500 then, which is wrong in HTTP/REST terms. You cannot set it to 401 for example when the exception is of type MyApp::Error::User::NotAuthorized.

**Comment by kraih** [comment]
I think you're misunderstanding exception handling in Mojolicious, it's not supposed to be the end all answer for handling every thrown exception. Quite the opposite actually, it is only meant to be a fallback for when your own exception handling fails, and/or during development. Throwing exceptions in controllers is not even encouraged, because of all the event loop features we support that are incompatible with that kind of code.

**Comment by flowdy** [comment]
I understand and will try on my own to catch and handle these errors properly, without relying on mojolicious, also to avoid it gets logged into the mojolicious log target. It is all about turning an app-specific exception, maybe thrown from deeply within the code, into a `my %error = (error => ref($e), message => $e->message, status => $e->http_status); $c->render( $c->stash('is_restapi_req') ? json => \%error : template => 'exception.production', %error ); return # from caller`, right?

I still think exactly this I would just prefer to define in a kind of "render_weak_exception" hook that could be called implicitly in

    use Try::Tiny;
    try {
        ...
    } catch { $c->reply->exception($_); 0 } or return;

Depending on whether my render_weak_exception hook returns a non-empty list, this is passed as is to `$c->render()` and that call simply returns without having upper levels of Mojolicious error handling even take notice. Otherwise it can do what it currently does, i.e. lead to a black error message and stacktrace in development mode or to an Internal Server Error and a log line in production mode.

tl;dr – In my opinion, Mojolicious could provide a comfortable means to deescalate exception objects thrown from deeper levels of code, to turn them to mere kind-of-normal responses with a certain http status code.
